WIP

AlexsanderDamaceno opened the pull request:
#### 3cca8e5b38f2c5ed24e5c389d225270946f8e253
<pre>
WIP
<a href="https://bugs.webkit.org/show_bug.cgi?id=300735">https://bugs.webkit.org/show_bug.cgi?id=300735</a>
Reviewed by NOBODY (OOPS!).

Instead of relying on a set of preprocessor checks to determine whether the execinfo.h header
and the backtrace/backtrace_symbols functions are available, it would be better to detect this with CMake.

* Source/WTF/wtf/PlatformHave.h:
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
* Source/cmake/FindExecinfo.cmake:
* Source/cmake/OptionsGTK.cmake:
* Source/WTF/wtf/PlatformGTK.cmake
* Source/WTF/wtf/PlatformWPE.cmake
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3cca8e5b38f2c5ed24e5c389d225270946f8e253

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132225 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4718 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139740 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84154 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4479 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101070 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68386 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135171 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3619 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118460 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82127 "Found 2 new API test failures: TestWTF:StackTraceTest.StackTraceWorks, TestWTF:StackTraceTest.CaptureStackTraceLimitAndSkip (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3504 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1098 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82960 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124288 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112398 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36579 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142387 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130732 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4387 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37159 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109447 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4468 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3796 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109628 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3324 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114732 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57634 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4441 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33105 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163699 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4273 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67887 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42538 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4532 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4400 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -28,6 +28,10 @@ find_package(ATSPI 2.5.3)
+find_package(Execinfo)
+SET_AND_EXPOSE_TO_BUILD(HAVE_BACKTRACE_FUNCTION ${Execinfo_FOUND})
+
+
```
Please be attentive to the whitespace expectations. Having two blank lines here is weird; the surrounding code does not do that:

```suggestion
```

aperezdc replied:
@mcatanzaro That's only true for glibc. There _is_ a `libexecinfo` which some Musl systems ship, and FreeBSD has such a library, too. So no, it is not “weird” to check for a library that we may need to be able to use `backtrace()`. And we want the imported target to refer to the library when available.

mcatanzaro replied:
Hm, OK, then using an imported target really does make sense, and my suggestion to put this in OptionsCommon.cmake doesn't work.

I also see that I was wrong about the imported target not doing anything. It actually does link in the libexecinfo library if `NOT HAVE_BACKTRACE AND HAVE_LIBEXECINFO`. That actually makes sense, but it's confusing and needs to be documented.

We need some more changes:

 * We need to make sure it's not an automagic dependency. The build should fail if libexecinfo is not found. You can add a new build option to allow building without libexecinfo if desired.
 * The imported target should itself provide the required definitions, and not define things globally.
 * The find module should document how it works, including all variables it defines and the target itself. Undefine anything that doesn't make sense to document.

aperezdc replied:
Provided that `backtrace()` and `backtrace_symbols()` may be provided by the libc _or_ alternatively by libexecinfo, in this particular case, and as an exception, I lean towards handling this automagically. The same way that linking librt is handled automagically.

Actually using `target_compile_definitions(Execinfo::Execinfo INTERFACE ...)` is a nice idea, and it does sound better than adding them where needed in different CMake files. It feels a bit awkward to have a `PAS_` prefixed definition in the `Execinfo::Execinfo` target, but I can live with that.

I agree that documenting the find-module would be good (@AlexsanderDamaceno: you can look at other modules to see how to document them).

mcatanzaro reviewed (changes requested):
Hm, this is part of glibc, not a distinct library, so it's a bit of a stretch to pretend that it is a library. Creating `HAVE_LIBEXECINFO` is weird because there is no such thing as libexecinfo. Same for using it as an imported target: it's really weird that an imported target would only add a couple preprocessor definitions rather than link to an actual library.

Actually, it's not the imported target that is carrying the preprocessor definitions. Those are being defined globally by the find module, which is surely not OK. I think your target is actually not doing _anything_, right?

You really don't need a find module or an imported target here. Just do the checks and add the definitions in OptionsCommon.cmake instead.

mcatanzaro commented on Source/cmake/FindExecinfo.cmake:
```diff
@@ -0,0 +1,48 @@
+if (HAVE_BACKTRACE)
+    add_definitions(-DHAVE_BACKTRACE_FUCTION=1)
+    add_definitions(-DPAS_HAVE_BACKTRACE_FUNCTION=1)
+endif()
```
These definitions are notably not used anywhere. There's no other reference to `BACKTRACE_FUNCTION` anywhere in WebKit. So I think you're missing some pieces here. You should add these in the same commit in which you start using them, not as a separate step.

aperezdc replied:
Agreed, this whole `if`-block can be removed. There is already a `SET_AND_EXPOSE_TO_BUILD()` for `HAVE_BACKTRACE` in `OptionsGTK.cmake`, which is what we prefer to generate the `cmakeconfig.h` header. We want to use `SET_AND_EXPOSE_TO_BUILD()` also for `HAVE_BACKTRACE_SYMBOLS`.

The guards in `Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c` will have to be changed to something like `#if PAS_OS(DARWIN) || PAS_HAVE(BACKTRACE)`, and the definition of `PAS_HAVE_BACKTRACE` should be added in the `bmalloc_DEFINITIONS` variable (file `Source/bmalloc/libpas/CMakeLists.txt`).

aperezdc replied:
On second thought, I think adding the definitions to the `Execinfo::Execinfo` target may be nicer in this case (see [my comment above](https://github.com/WebKit/WebKit/pull/53431#discussion_r2496443584)).

aperezdc commented on Source/cmake/FindExecinfo.cmake:
```diff
@@ -0,0 +1,48 @@
+    find_library(EXECINFO_LIBRARY execinfo)
+    if (EXECINFO_LIBRARY)
+        set(CMAKE_REQUIRED_LIBRARIES ${EXECINFO_LIBRARY})
+        check_function_exists(backtrace HAVE_BACKTRACE)
```
Here we want to use `cmake_push_check_state(RESET)` to start with clean variables for the checks, and then restore the previous state with `cmake_pop_check_state()`:


```suggestion
        cmake_push_check_state(RESET)
        set(CMAKE_REQUIRED_LIBRARIES ${EXECINFO_LIBRARY})
        check_function_exists(backtrace HAVE_BACKTRACE)
        cmake_pop_check_state()
```

aperezdc commented on Source/cmake/FindExecinfo.cmake:
```diff
@@ -0,0 +1,48 @@
+check_function_exists(backtrace HAVE_BACKTRACE)
+
+if (NOT HAVE_BACKTRACE)
+    find_library(EXECINFO_LIBRARY execinfo)
```
Variables for find-modules typically have the name of the module with the same capitalization, so this should be `Execinfo_LIBRARY`:


```suggestion
    find_library(Execinfo_LIBRARY execinfo)
```

aperezdc commented on Source/cmake/FindExecinfo.cmake:
```diff
@@ -0,0 +1,48 @@
+    endif()
+endif()
+
+if (HAVE_LIBEXECINFO)
```
This needs to check that the target was not already created:


```suggestion
if (Execinfo_LIBRARY AND NOT TARGET Execinfo::Execinfo)
```

aperezdc commented on Source/cmake/FindExecinfo.cmake:
```diff
@@ -0,0 +1,48 @@
+# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+check_include_file("execinfo.h" HAVE_EXECINFO_H)
+check_function_exists(backtrace HAVE_BACKTRACE)
```
We would also want to check for the `backtrace_symbols()` function here.

aperezdc commented on Source/cmake/FindExecinfo.cmake:
```diff
@@ -0,0 +1,48 @@
+        set(CMAKE_REQUIRED_LIBRARIES ${EXECINFO_LIBRARY})
+        check_function_exists(backtrace HAVE_BACKTRACE)
+        if (HAVE_BACKTRACE)
+            set(HAVE_LIBEXECINFO 1)
```
The convention for find-modules is to use `<name>_FOUND`, so this should be `Execinfo_FOUND`.


```suggestion
            set(Execinfo_FOUND TRUE)
```

aperezdc commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -28,6 +28,10 @@ find_package(ATSPI 2.5.3)
 include(FindGLibCompileResources)
 
+find_package(Execinfo)
+SET_AND_EXPOSE_TO_BUILD(HAVE_BACKTRACE_FUNCTION ${Execinfo_FOUND})
```
This can be just `HAVE_BACKTRACE`. But you need to edit `Source/WTF/wtf/PlatformHave.h` to change the definition in that file. Currently it is defined depending on preprocessor guards:

```cpp
#if OS(DARWIN) || OS(HAIKU) || ((OS(LINUX) || OS(HURD)) && defined(__GLIBC__) && !defined(__UCLIBC__) && !CPU(MIPS))
#define HAVE_BACKTRACE 1
#define HAVE_BACKTRACE_SYMBOLS 1
#endif
```

We will want something like this:

```cpp
#if OS(DARWIN)
#ifndef HAVE_BACKTRACE
#define HAVE_BACKTRACE 1
#endif
#ifndef HAVE_BACKTRACE_SYMBOLS
#define HAVE_BACKTRACE_SYMBOLS 1
#endif
#endif
```

This is because the Apple ports (= Darwin) do not use CMake, so the checks we add in CMake are not used for them.

aperezdc commented on Source/bmalloc/CMakeLists.txt:
This won't work, instead add the target name, if defined, to the `bmalloc_PRIVATE_LIBRARIES` variable:


```suggestion
    if (TARGET Execinfo::Execinfo)
        list(APPEND bmalloc_PRIVATE_LIBRARIES Execinfo::Execinfo)
    endif ()
```

This is the same you did for `WTF_PRIVATE_LIBRARIES` below.

aperezdc commented on Source/cmake/FindExecinfo.cmake:
This will always fail, CMake variable names are case-sensitive:

```suggestion
    if (Execinfo_LIBRARY)
        cmake_push_check_state(RESET)
        set(CMAKE_REQUIRED_LIBRARIES ${Execinfo_LIBRARY})
```

aperezdc commented on Source/cmake/FindExecinfo.cmake:
This should also check whether the found libexecinfo also has the `backtrace_symbols()` function. Please add the missing check.

aperezdc reviewed (changes requested):
This still needs a couple of fixes, please take a look at the comments.

aperezdc commented on Source/cmake/FindExecinfo.cmake:
Instead of manually setting `Execinfo_FOUND`, we can use the `find_package_handle_standard_args()` helper function:


```suggestion
    endif ()
endif ()

include(FindPackageHandleStandardArgs)
find_package_handle_standard_args(Execinfo
    REQUIRED_VARS HAVE_EXECINFO_H HAVE_BACKTRACE
)
```

This will also produce a status message with the usual format, and check that both the `execinfo.h` header and the `backtrace()` function are present.

aperezdc commented on Source/cmake/FindExecinfo.cmake:
This defines the `HAVE_EXECINFO_H` variable, but then it is not used anywhere. See my suggestion above to use it with `find_package_handle_standard_args()`.

aperezdc reviewed (changes requested):
I have one last comment. After fixing that and adding documentation to the find-module we can merge this.

AlexsanderDamaceno committed:
WIP
https://bugs.webkit.org/show_bug.cgi?id=300735
Reviewed by NOBODY (OOPS!).

Instead of relying on a set of preprocessor checks to determine whether the execinfo.h header
and the backtrace/backtrace_symbols functions are available, it would be better to detect this with CMake.

* Source/WTF/wtf/PlatformHave.h:
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
* Source/cmake/FindExecinfo.cmake:
* Source/cmake/OptionsGTK.cmake:
* Source/WTF/wtf/PlatformGTK.cmake
* Source/WTF/wtf/PlatformWPE.cmake
